File: app/src/main/java/br/com/popularmovies/moviedetail/viewmodel/MovieDetailViewModel.java
package br.com.popularmovies.moviedetail.viewmodel;

import androidx.arch.core.util.Function;
import androidx.lifecycle.LiveData;
import androidx.lifecycle.MutableLiveData;
import androidx.lifecycle.Transformations;
import androidx.lifecycle.ViewModel;

import br.com.popularmovies.data.model.Resource;
import br.com.popularmovies.services.movieService.response.Movie;
import br.com.popularmovies.services.movieService.source.MovieRepository;

public class MovieDetailViewModel extends ViewModel {
    private final LiveData<Resource<Void>> favorites;
    private final LiveData<Resource<Movie>> mMovie;
    private final MutableLiveData<Boolean> movieStatus = new MutableLiveData<>();
    private final MutableLiveData<Integer> _movieId = new MutableLiveData<>();
    private Movie movie;

    public MovieDetailViewModel(final MovieRepository mMovieRepository, final int movieId) {
        mMovie = Transformations.switchMap(_movieId, new Function<Integer, LiveData<Resource<Movie>>>() {
            @Override
            public LiveData<Resource<Movie>> apply(Integer input) {
                return mMovieRepository.getMovie(movieId);
            }
        });
        _movieId.setValue(movieId);
        favorites = Transformations.switchMap(movieStatus, new Function<Boolean, LiveData<Resource<Void>>>() {
            @Override
            public LiveData<Resource<Void>> apply(Boolean isFavorite) {
                if (isFavorite != null) {
                    movie.setFavorite(isFavorite);
                    if (isFavorite) {
                        return mMovieRepository.saveMovie(movie);
                    } else {
                        return mMovieRepository.removeMovie(movie);
                    }
                }
                return null;
            }
        });
    }


    public LiveData<Resource<Void>> getFavorites() {
        return favorites;
    }

    public LiveData<Resource<Movie>> getmMovie() {
        return mMovie;
    }

    public Movie getMovie() {
        return movie;
    }

    public void setMovie(Movie movie) {
        this.movie = movie;
    }

    public void saveFavorites(boolean status) {
        this.movieStatus.setValue(status);
    }

    public void tryAgain() {
        _movieId.setValue(_movieId.getValue());
    }
}

File: app/src/main/java/br/com/popularmovies/moviedetail/viewmodel/MovieDetailViewModel.kt
package br.com.popularmovies.moviedetail.viewmodel

import androidx.lifecycle.LiveData
import androidx.lifecycle.MutableLiveData
import androidx.lifecycle.ViewModel
import androidx.lifecycle.viewModelScope
import br.com.popularmovies.common.models.base.NetworkError
import br.com.popularmovies.common.models.base.Result
import br.com.popularmovies.model.movie.Movie
import br.com.popularmovies.model.movie.MovieReview
import br.com.popularmovies.model.movie.MovieTrailer
import br.com.popularmovies.domain.usecases.movies.GetMovieUseCase
import br.com.popularmovies.domain.usecases.movies.reviews.GetMovieReviewsUseCase
import br.com.popularmovies.domain.usecases.movies.trailers.GetMovieTrailersUseCase
import dagger.assisted.Assisted
import dagger.assisted.AssistedFactory
import dagger.assisted.AssistedInject
import kotlinx.coroutines.launch

class MovieDetailViewModel @AssistedInject constructor(
    private val getMovieUseCase: GetMovieUseCase,
    private val getMovieTrailersUseCase: GetMovieTrailersUseCase,
    private val getMovieReviewsUseCase: GetMovieReviewsUseCase,
    @Assisted private val movieId: Long
) : ViewModel() {

    @AssistedFactory
    interface Factory {
        fun create(movieId: Long): MovieDetailViewModel
    }

    val loading = MutableLiveData<Boolean>()

    private val _error = MutableLiveData<NetworkError>()
    val error: LiveData<NetworkError>
        get() = _error

    private val _movie = MutableLiveData<Movie>()
    val movie: LiveData<Movie>
        get() = _movie

    private val _trailers = MutableLiveData<List<MovieTrailer>>()
    val trailers: LiveData<List<MovieTrailer>> = _trailers

    private val _reviews = MutableLiveData<List<MovieReview>>()
    val reviews: LiveData<List<MovieReview>> = _reviews

    private val _playTrailer = MutableLiveData<String>()
    val playTrailer: LiveData<String>
        get() = _playTrailer

    init {
        getMovie()
        getTrailers()
        getReviews()
    }

    private fun getMovie() {
        viewModelScope.launch {
            showLoading(true)
            val params = GetMovieUseCase.Params(movieId)
            when (val result = getMovieUseCase.build(params)) {
                is Result.Success -> {
                    _movie.value = result.data
                }

                is Result.Error -> _error.value = result.error
            }
        }
    }

    private fun getTrailers() {
        viewModelScope.launch {
            showLoading(true)
            val params = GetMovieTrailersUseCase.Params(movieId)
            when (val result = getMovieTrailersUseCase.build(params)) {
                is Result.Success -> _trailers.value = result.data
                is Result.Error -> _error.value = result.error
            }
        }
    }

    private fun getReviews() {
        viewModelScope.launch {
            showLoading(true)
            val params = GetMovieReviewsUseCase.Params(movieId)
            when (val result = getMovieReviewsUseCase.build(params)) {
                is Result.Success -> _reviews.value = result.data
                is Result.Error -> _error.value = result.error
            }
        }
    }

    fun showLoading(value: Boolean) {
        loading.value = value
    }

    fun playTrailer(key: String) {
        _playTrailer.value = key
    }
}